File: frontend/src/types/analysis.ts
export type Trend = 'bullish' | 'bearish' | 'ranging';
export type Phase = 'compression' | 'expansion' | 'normal';

export interface SwingPoint {
  time: number;
  price: number;
  kind: 'high' | 'low';
  strength: 'major' | 'minor';
  label: 'HH' | 'HL' | 'LH' | 'LL' | null;
}

export interface Level {
  price: number;
  kind: 'support' | 'resistance';
  touches: number;
  first_time: number;
  last_time: number;
}

export interface LinePoint {
  time: number;
  price: number;
}

export interface Trendline {
  start: LinePoint;
  end: LinePoint;
  kind: 'support' | 'resistance';
  touches: number;
  score: number;
}

export interface StructureEvent {
  type: 'BOS' | 'CHoCH';
  direction: 'bullish' | 'bearish';
  price: number;
  from_time: number;
  time: number;
}

export interface Liquidity {
  price: number;
  side: 'buy' | 'sell';
  touches: number;
}

export interface Zone {
  kind: 'order_block' | 'fvg';
  direction: 'bullish' | 'bearish';
  top: number;
  bottom: number;
  start_time: number;
  end_time: number;
  mitigated: boolean;
  fresh: boolean;
  tests: number;
  strength: number;
  label: string;
}

export interface FibLevel {
  ratio: number;
  price: number;
}

export interface Fibonacci {
  high: number;
  low: number;
  direction: 'up' | 'down';
  start_time: number;
  end_time: number;
  levels: FibLevel[];
}

export interface KeyLevel {
  label: string;
  price: number;
}

export interface CandleMark {
  time: number;
  name: string;
  bias: 'bullish' | 'bearish' | 'neutral';
}

export interface StructureResponse {
  symbol: string;
  interval: string;
  trend: Trend;
  trend_strength: number;
  phase: Phase;
  swings: SwingPoint[];
  levels: Level[];
  trendlines: Trendline[];
  events: StructureEvent[];
  liquidity: Liquidity[];
  zones: Zone[];
  fibonacci: Fibonacci | null;
  key_levels: KeyLevel[];
  patterns: CandleMark[];
}

// --- Phase 4: decision / confidence / risk ---

export type Decision = 'BUY' | 'SELL' | 'WAIT' | 'NO_TRADE';

export interface CandlePattern {
  name: string;
  bias: 'bullish' | 'bearish' | 'neutral';
}

export interface DecisionScores {
  trend: number;
  structure: number;
  momentum: number;
  pattern: number;
  risk: number;
  volume: number;
}

export interface RiskPlan {
  direction: 'long' | 'short';
  entry: number;
  stop_loss: number;
  take_profit: number;
  risk_reward: number;
  risk_pct: number;
  account_balance: number;
  position_size: number;
}

export interface SignalResponse {
  symbol: string;
  interval: string;
  decision: Decision;
  bias: Trend;
  confidence: number;
  trend: Trend;
  momentum_rsi: number | null;
  scores: DecisionScores;
  pattern: CandlePattern | null;
  risk: RiskPlan | null;
  reasons: string[];
}

export interface MentorResponse {
  symbol: string;
  interval: string;
  configured: boolean;
  explanation: string;
  decision: string | null;
  confidence: number | null;
}

export interface ScanRow {
  symbol: string;
  name: string;
  type: string;
  price: number | null;
  decision: Decision;
  bias: Trend;
  confidence: number;
  trend: Trend;
  risk_reward: number | null;
  rsi: number | null;
}

export interface ScanResponse {
  interval: string;
  rows: ScanRow[];
}

// --- Module 17: Backtesting ---

export interface BacktestTrade {
  entry_time: number;
  exit_time: number;
  direction: 'long' | 'short';
  entry: number;
  stop_loss: number;
  take_profit: number;
  exit_price: number;
  outcome: 'win' | 'loss' | 'timeout';
  r_multiple: number;
  pnl: number;
  confidence: number;
}

export interface EquityPoint {
  time: number;
  balance: number;
}

export interface BacktestStats {
  total_trades: number;
  wins: number;
  losses: number;
  win_rate: number;
  profit_factor: number | null;
  expectancy_r: number;
  avg_win_r: number;
  avg_loss_r: number;
  avg_rr: number;
  total_return_pct: number;
  final_balance: number;
  max_drawdown_pct: number;
}

export interface BacktestResponse {
  symbol: string;
  interval: string;
  account_balance: number;
  risk_pct: number;
  min_confidence: number;
  candles_tested: number;
  stats: BacktestStats;
  trades: BacktestTrade[];
  equity_curve: EquityPoint[];
}

File: backend/app/schemas/analysis.py
"""Pydantic models for the analysis / drawing engine (Phase 2)."""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class SwingPoint(BaseModel):
    time: int = Field(..., description="UNIX epoch seconds of the pivot candle")
    price: float
    kind: Literal["high", "low"]
    strength: Literal["major", "minor"] = "minor"
    label: Literal["HH", "HL", "LH", "LL"] | None = None


class Level(BaseModel):
    """A horizontal support/resistance level clustered from swing points."""

    price: float
    kind: Literal["support", "resistance"]
    touches: int = Field(..., description="Number of swing points that formed this level")
    first_time: int
    last_time: int


class LinePoint(BaseModel):
    time: int
    price: float


class Trendline(BaseModel):
    start: LinePoint
    end: LinePoint
    kind: Literal["support", "resistance"]
    touches: int
    score: float = Field(..., description="Higher = stronger (more touches, better fit)")


class StructureEvent(BaseModel):
    """A Break of Structure (BOS) or Change of Character (CHoCH)."""

    type: Literal["BOS", "CHoCH"]
    direction: Literal["bullish", "bearish"]
    price: float = Field(..., description="The broken swing level")
    from_time: int = Field(..., description="Time of the swing that was broken")
    time: int = Field(..., description="Time of the candle that broke it")


class Liquidity(BaseModel):
    """A liquidity pool from equal highs/lows."""

    price: float
    side: Literal["buy", "sell"]  # buy-side rests above price, sell-side below
    touches: int


class Zone(BaseModel):
    """A supply/demand zone: Order Block or Fair Value Gap (from SMC)."""

    kind: Literal["order_block", "fvg"]
    direction: Literal["bullish", "bearish"]
    top: float
    bottom: float
    start_time: int
    end_time: int
    mitigated: bool = False
    fresh: bool = True
    tests: int = 0
    strength: int = 50
    label: str = ""


Trend = Literal["bullish", "bearish", "ranging"]
Phase = Literal["compression", "expansion", "normal"]


class FibLevel(BaseModel):
    ratio: float
    price: float


class Fibonacci(BaseModel):
    high: float
    low: float
    direction: Literal["up", "down"]
    start_time: int
    end_time: int
    levels: list[FibLevel]


class KeyLevel(BaseModel):
    label: str  # PDH, PDL, PWH, PWL
    price: float


class CandleMark(BaseModel):
    time: int
    name: str
    bias: Literal["bullish", "bearish", "neutral"]


class StructureResponse(BaseModel):
    symbol: str
    interval: str
    trend: Trend
    trend_strength: int = Field(..., ge=0, le=100)
    phase: Phase
    swings: list[SwingPoint]
    levels: list[Level]
    trendlines: list[Trendline]
    events: list[StructureEvent]
    liquidity: list[Liquidity]
    zones: list[Zone]
    fibonacci: Fibonacci | None = None
    key_levels: list[KeyLevel] = []
    patterns: list[CandleMark] = []


# --- Phase 4: Decision / Confidence / Risk ---------------------------------


class CandlePattern(BaseModel):
    name: str
    bias: Literal["bullish", "bearish", "neutral"]


class DecisionScores(BaseModel):
    trend: int
    structure: int
    momentum: int
    pattern: int
    risk: int
    volume: int


class RiskPlan(BaseModel):
    direction: Literal["long", "short"]
    entry: float
    stop_loss: float
    take_profit: float
    risk_reward: float
    risk_pct: float
    account_balance: float
    position_size: float = Field(..., description="Units sized to risk_pct of balance")


class SignalResponse(BaseModel):
    symbol: str
    interval: str
    decision: Literal["BUY", "SELL", "WAIT", "NO_TRADE"]
    bias: Trend
    confidence: int = Field(..., ge=0, le=100)
    trend: Trend
    momentum_rsi: float | None
    scores: DecisionScores
    pattern: CandlePattern | None
    risk: RiskPlan | None
    reasons: list[str]


# --- Phase 5: AI Mentor ---


class MentorResponse(BaseModel):
    symbol: str
    interval: str
    configured: bool = Field(..., description="Whether the AI mentor is enabled (API key set)")
    explanation: str = ""
    decision: str | None = None
    confidence: int | None = None


class ChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str = Field(..., max_length=4000)


class MentorChatRequest(BaseModel):
    symbol: str
    interval: str
    messages: list[ChatMessage] = Field(..., min_length=1, max_length=20)


# --- Module 14: Market Scanner ---


class ScanRow(BaseModel):
    symbol: str
    name: str
    type: str
    price: float | None = None
    decision: Literal["BUY", "SELL", "WAIT", "NO_TRADE"]
    bias: Trend
    confidence: int
    trend: Trend
    risk_reward: float | None = None
    rsi: float | None = None


class ScanResponse(BaseModel):
    interval: str
    rows: list[ScanRow]


# --- Module 15: Notifications ---


class AlertRequest(BaseModel):
    title: str = Field(..., max_length=120)
    message: str = Field(..., max_length=1000)


class NotifyStatus(BaseModel):
    discord: bool
    telegram: bool


class AlertResult(BaseModel):
    sent: list[str]


# --- Module 17: Backtesting ---


class BacktestTrade(BaseModel):
    """A single simulated trade produced by the walk-forward backtest."""

    entry_time: int
    exit_time: int
    direction: Literal["long", "short"]
    entry: float
    stop_loss: float
    take_profit: float
    exit_price: float
    outcome: Literal["win", "loss", "timeout"]
    r_multiple: float = Field(..., description="Result in R (reward:risk) units")
    pnl: float = Field(..., description="Profit/loss in account currency")
    confidence: int


class EquityPoint(BaseModel):
    time: int
    balance: float


class BacktestStats(BaseModel):
    total_trades: int
    wins: int
    losses: int
    win_rate: float = Field(..., description="Winners / total (%)")
    profit_factor: float | None = Field(None, description="Gross profit / gross loss")
    expectancy_r: float = Field(..., description="Average R per trade")
    avg_win_r: float
    avg_loss_r: float
    avg_rr: float = Field(..., description="Average planned reward:risk")
    total_return_pct: float
    final_balance: float
    max_drawdown_pct: float


class BacktestResponse(BaseModel):
    symbol: str
    interval: str
    account_balance: float
    risk_pct: float
    min_confidence: int
    candles_tested: int
    stats: BacktestStats
    trades: list[BacktestTrade]
    equity_curve: list[EquityPoint]
